File: src/entities/bacteria/Bacteria.java
package entities.bacteria;

import controllers.Main;
import controllers.ModificationsPane;
import entities.MCP.MCP;
import food.NanoFoodPiece;
import javafx.geometry.Point2D;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.Pane;
import javafx.scene.transform.Rotate;

import java.util.HashSet;

import static controllers.Main.gL;

public class Bacteria extends Pane{
    public static ImageView activeCircle = new ImageView(new Image("res/Active.png"));

    public HashSet<BMod> modifications = new HashSet<>();

    private ModificationsPane mp = new ModificationsPane();
    private Point2D position;
    private String caller = "idle";
    private Point2D direction = new Point2D(0, 0);
    private double currentVelocity = 1;
    private double velocity = 1;
    private float hungerThreshold = 10;
    private float satiety = 10;
    private float maxSatiety = 100;
    private NanoFoodPiece choose = new NanoFoodPiece();
    private StateMachine state = StateMachine.idle;
    public boolean remove = false;
    private Point2D nextPos;
    private ImageView sprite = new ImageView(new Image("res/Bacteria1.png"));

    private Bacteria partner;
    private boolean isWaitingForPartner;

    private float waitingT = 1;
    private float modInhChance = 20;
    private float[] nucleus = new float[4];

    public Bacteria(double x, double y, Bacteria parent) {
        sprite.setFitHeight(48);
        sprite.setFitWidth(69);
        sprite.setX(-(sprite.getFitWidth() / 2));
        sprite.setY(-(sprite.getFitHeight() / 2));

        modifications.add(BMod.None);

        activeCircle.setFitHeight(75);
        activeCircle.setFitWidth(75);
        activeCircle.setX(-activeCircle.getFitWidth() / 2);
        activeCircle.setY(-activeCircle.getFitWidth() / 2);
        setTranslateY(y);
        setTranslateX(x);

        this.setPrefHeight(sprite.getFitHeight());
        this.setPrefWidth(sprite.getFitWidth());
        getChildren().addAll(sprite);
        currentVelocity = velocity;

        inherit(parent);

        sprite.setOnMouseClicked(event -> {
            if (Main.BTarget != this) {
                Main.BTarget = this;
                Main.bMap.forEach(next -> {
                    if (next.getChildren().contains(activeCircle)) {
                        next.getChildren().remove(activeCircle);
                    }
                });
                getChildren().add(activeCircle);
            } else {
                Main.BTarget = null;
                getChildren().remove(activeCircle);
            }
        });
    }

    //     Every tick event
    public void update(HashSet<NanoFoodPiece> foodPeaces) {
        if (Main.BTarget != this && getChildren().contains(activeCircle)) getChildren().remove(activeCircle);
        satiety -= .006;
        Point2D front = new Point2D(getTranslateX(), getTranslateY()).add(direction.multiply(sprite.getFitWidth() / 2));
        switch (state) {
            case isWait:
                waitingT--;
                if (waitingT == 0) {
                    switch (caller) {
                        case "idle":
                            changeState("idle");
                    }
                }
                break;
            case isRotating:
                break;
            case isMoving:
                if (caller.equals("fs") && !Main.root.getChildren().contains(choose)) searchForFood(Main.foodList);
                if (front.distance(nextPos) < front.distance(front.add(direction.multiply(currentVelocity)))) {
                    switch (caller) {
                        case "fs":
                            eats();
                            caller = "idle";
                            break;
                        case "idle":
                            changeState("idle");
                            break;
                        case "mcp":
                            if (partner.isWaitingForPartner) {
                                bornMcp();
                            } else isWaitingForPartner = true;
                            break;
                    }
                    setTranslateY(getTranslateY() + (nextPos.getY() - front.getY()));
                    setTranslateX(getTranslateX() + (nextPos.getX() - front.getX()));
                } else {
                    setTranslateY(getTranslateY() + direction.getY() * currentVelocity);
                    setTranslateX(getTranslateX() + direction.getX() * currentVelocity);
                }

                break;
            case idle:
                currentVelocity = velocity / 4;
                if (satiety < hungerThreshold) {
                    searchForFood(foodPeaces);
                }
                moveToPoint(new Point2D(getTranslateX() + Math.random() * 1400 - 700, getTranslateY() + Math.random() * 1400 - 700));
                break;
        }
        position = new Point2D(getTranslateX(), getTranslateY());
    }

    private void moveToPoint(Point2D pos) {
        //nextPos is using on the method outside(usable)
        nextPos = pos;
        direction = pos.subtract(getTranslateX(), getTranslateY()).normalize();
        double angle = calculateAngle(direction.getX(), direction.getY());
        getTransforms().clear();
        Rotate rotator = new Rotate(angle);
        getTransforms().add(rotator);
        changeState("isMoving");
    }

    public boolean canMultiply() {
        return satiety >= 20;
    }

    public void updateInhInfo() {
        nucleus[0] = satiety;
        nucleus[1] = maxSatiety;
        nucleus[2] = hungerThreshold;
        nucleus[3] = modInhChance;
    }

    private void inherit(Bacteria parent) {
        parent.modifications.forEach(next -> {
            float[] params = parent.nucleus;
            if (Math.random() * 100 <= params[3]) {
                addModification(next);
            }
            satiety = (float) (Math.random() * 100 <= 40 ? params[0] + Math.random() * 2 - 1 : params[0]);
            maxSatiety = (float) (Math.random() * 100 <= 40 ? params[1] + Math.random() * 2 - 1 : params[1]);
            hungerThreshold = (float) (Math.random() * 100 <= 40 ? params[2] + Math.random() * 2 - 1 : params[2]);
            modInhChance = (float) (Math.random() * 100 <= 40 ? params[3] + Math.random() * 2 - 1 : params[3]);
        });
    }

    public void addModification(BMod mod) {
        if (!mod.equals(BMod.None)) {
            modifications.add(mod);
            getChildren().add(mod);
        }
        mod.setX(mod.x);
        mod.setY(mod.y);
        modifications.remove(BMod.None);
        satiety -= mod.price;
        if (mod.isOutside) mod.toBack();
    }

    public void showModPane() {
        mp.reinitializeAndShow();
    }


    //    angle between x axis of bacteria and point with signature declared coordinates
    private double calculateAngle(double vexX, double vecY) {
        double angle = new Point2D(vexX, vecY).angle(1, 0);
        return vecY > 0 ? angle : -angle;
    }

    private void changeState(String state) {
        this.state = StateMachine.valueOf(state);
    }

    public void searchForFood(HashSet<NanoFoodPiece> foodList) {
        if (satiety >= maxSatiety - 1) {
            Main.printMessage("entities.bacteria.Bacteria is full");
            return;
        }
        double previous = 1000;
//        Choosing
        Point2D selfPosition = new Point2D(getTranslateX(), getTranslateY());
        for (NanoFoodPiece next : foodList) {
            if (new Point2D(next.getTranslateX(), next.getTranslateY()).distance(selfPosition) < previous) {
                previous = selfPosition.distance(new Point2D(next.getTranslateX(), next.getTranslateY()));
                choose = next;
            }
        }
//        Checking
        if (choose == null) {
            Main.printMessage("food not founded");
        } else {
            caller = "fs";
            currentVelocity = velocity;
            moveToPoint(choose.getPosition());
        }
    }

    private void eats() {
        satiety = (float) (satiety + choose.getFoodValue() > maxSatiety ? maxSatiety : satiety + choose.getFoodValue() * 10);
        choose.destroy();
        waitingT = 120;
        changeState("isWait");
        caller = "idle";
    }

    public Bacteria searchForPartner(HashSet<Bacteria> bList) {
        Bacteria relatePartner = null;
        double previous = 1000;
//        searching
        for (Bacteria next : bList) {
            if (next.position.distance(position) < previous && next != this) {
                previous = position.distance(next.position);
                relatePartner = next;
            }
        }
//        checking
        currentVelocity = velocity;
        return relatePartner;
    }

    public void makeMCP(Bacteria right, Bacteria left) {
        try {
            partner = left;
            partner.partner = right;
            if (right.position.distance(left.position) > 50) {
                moveToPoint(right.position.midpoint(left.position));
                left.moveToPoint(right.position.midpoint(left.position));
                partner.caller = "mcp";
                caller = "mcp";
            } else bornMcp();
        } catch (NullPointerException e) {
            Main.printMessage("Partner not founded");
        }
    }

    private void bornMcp() {
        MCP nmcp = new MCP();
        nmcp.setTranslateX(Main.BTarget.getTranslateX());
        nmcp.setTranslateY(Main.BTarget.getTranslateY());
        Main.mcpMap.add(nmcp);
        Main.root.getChildren().add(nmcp);
        if (Main.BTarget == this || Main.BTarget == partner) {
            destroy();
            partner.destroy();
            Main.BTarget = null;
            Main.MCPTarget = nmcp;
        } else {
            destroy();
            partner.destroy();
        }
    }

    public Bacteria(double x, double y) {
        sprite.setFitHeight(191 * .25);
        sprite.setFitWidth(276 * .25);
        sprite.setX(-(sprite.getFitWidth() / 2));
        sprite.setY(-(sprite.getFitHeight() / 2));
        modifications.add(BMod.None);
        setTranslateX(x);
        setTranslateY(y);

        activeCircle.setFitHeight(75);
        activeCircle.setFitWidth(75);
        activeCircle.setX(-activeCircle.getFitWidth() / 2);
        activeCircle.setY(-activeCircle.getFitWidth() / 2);

        this.setPrefHeight(sprite.getFitHeight());
        this.setPrefWidth(sprite.getFitWidth());
        getChildren().addAll(sprite);
        currentVelocity = velocity;
        sprite.setOnMouseClicked(event -> {
            if (Main.BTarget != this) {
                Main.BTarget = this;
                Main.bMap.forEach(next -> {
                    if (next.getChildren().contains(activeCircle)) {
                        next.getChildren().remove(activeCircle);
                    }
                });
                getChildren().add(activeCircle);
            } else {
                Main.BTarget = null;
                getChildren().remove(activeCircle);
            }
        });
    }

    private void destroy() {
        Main.root.getChildren().remove(this);
        remove = true;
    }

    @Override
    public String toString() {
        String beforeMods = gL("target", "Target") + ":\n" + gL("satiety", "Satiety") + " - " + Math.round(satiety);
        StringBuilder mods = new StringBuilder("\n" + gL("modifications", "Modifications"));
        modifications.forEach(mod -> mods.append("\n-").append(mod.name));
        return beforeMods + mods;
    }

    public float getSatiety() {
        return satiety;
    }

    public void setSatiety(float satiety) {
        this.satiety = satiety;
    }
}

File: src/main/kotlin/shmax/entity/bacteria/Bacteria.kt
package shmax.entity.bacteria

import javafx.event.EventHandler
import javafx.geometry.Point2D
import javafx.scene.image.ImageView
import javafx.scene.layout.Pane
import javafx.scene.transform.Rotate
import shmax.controllers.MainLoop
import shmax.controllers.MainLoop.Companion.gL
import shmax.entity.OrganismState
import shmax.entity.cellular.MultiCellularOrganism
import shmax.food.NanoFoodPiece
import shmax.generated.R
import shmax.util.*
import kotlin.math.min
import kotlin.math.round
import kotlin.random.Random

class Bacteria(
    x: Double,
    y: Double,
    parent: Bacteria? = null
): Pane() {
    companion object {
        val activeCircle = ImageView(R.images.sprite_active).apply {
            fitHeight = 75.0
            fitWidth = 75.0
        }
        const val GRAPHICS_HEIGHT = 48.0
        const val GRAPHICS_WIDTH = 69.0
    }

    val modifications = mutableSetOf<BacteriaModification>()

    private var position = Point2D(.0, .0)
    private var direction = Point2D(.0, .0)
    private var currentVelocity = 1.0
    private val baseVelocity = 1.0
    private var hungerThreshold = 10f
    var satiety = 10f
    private var maxSatiety = 100f
    val canMultiply
        get() = satiety >= 20f
    private var foodTarget = NanoFoodPiece()
    private var state = OrganismState.IDLE
    var remove = false
    private var movingTarget = Point2D(0.0, 0.0)
    private val graphics = ImageView(R.images.sprite_bacteria_bacteria1)

    private var caller = "idle"

    private var partnerTarget: Bacteria? = null
    private var waitingForPartner = false
    private var waitingTime = 1f
    private var modificationInheritanceChance = 20f

    private var nucleus = FloatArray(4)


    init {
        graphics.fitWidth = GRAPHICS_WIDTH
        graphics.fitHeight = GRAPHICS_HEIGHT

        graphics.x = -GRAPHICS_WIDTH / 2
        graphics.y = -GRAPHICS_HEIGHT / 2

        modifications += BacteriaModification(BacteriaModificationDef.NONE)

        activeCircle.x = -activeCircle.fitWidth / 2
        activeCircle.y = -activeCircle.fitHeight / 2

        translatePosition = Point2D(x, y)

        prefHeight = graphics.fitHeight
        prefWidth = graphics.fitWidth

        children.add(graphics)
        currentVelocity = baseVelocity

        inherit(parent)

        graphics.onMouseClicked = EventHandler {
            if (MainLoop.bacteriaTarget != this) {
                MainLoop.bacteriaTarget = this
                MainLoop.bacteriaList.forEach { bacteria ->
                    bacteria.children.remove(activeCircle)
                }
                children.add(activeCircle)
            } else {
                MainLoop.bacteriaTarget = null
                children.remove(activeCircle)
            }
        }
    }

    fun update(foodPeaces: Set<NanoFoodPiece>) {
        if (MainLoop.bacteriaTarget != this) {
            children.remove(activeCircle)
        }

        satiety -= .006f

        val frontVector = translatePosition + (direction * (graphics.fitWidth / 2))

        when (state) {
            OrganismState.IDLE -> {
                currentVelocity = baseVelocity / 4
                if (satiety < hungerThreshold) {
                    searchForFood(foodPeaces)
                }
                moveToPoint(translatePosition + Random.nextDouble(700.0, 1400.0))
            }
            OrganismState.MOVING -> {
                if (caller == "fs" && foodTarget !in MainLoop.objectRoot.children) {
                    searchForFood(MainLoop.foodList)
                }
                if (frontVector.distance(movingTarget) < frontVector.distance(frontVector + (direction * currentVelocity))) {
                    when (caller) {
                        "fs" -> {
                            eatFoodTarget()
                            caller = "idle"
                        }
                        "mcp" -> if (partnerTarget?.waitingForPartner == true) {
                            bornMultiCellularOrganism()
                        } else waitingForPartner = true
                    }
                    translatePosition += (movingTarget - frontVector)
                } else {
                    translatePosition += (direction * currentVelocity)
                }
            }
            OrganismState.ROTATING -> {}
            OrganismState.WAITING -> {
                waitingTime--
                if (waitingTime <= 0 && caller == "idle") {
                    state = OrganismState.IDLE
                }
            }
        }

        position = translatePosition
    }

    private fun moveToPoint(point: Point2D) {
        movingTarget = point
        direction = (point - translatePosition).normalize()
        val angle = calculateAngle(direction)

        transforms.clear()
        transforms.add(Rotate(angle))
        state = OrganismState.MOVING
    }

    private fun calculateAngle(point: Point2D): Double {
        val angle = point.angle(1.0, 0.0)
        return if (point.y > 0) angle else -angle
    }

    fun updateInheritanceInfo() {
        nucleus[0] = satiety
        nucleus[1] = maxSatiety
        nucleus[2] = hungerThreshold
        nucleus[3] = modificationInheritanceChance
    }

    private fun inherit(parent: Bacteria?) {
        parent ?: return
        parent.modifications.forEach { modification ->
            val params = parent.nucleus
            if (rollChance(100) <= params[3]) {
                addModification(modification)
            }
            val randomizeParam = { Random.nextInt(1, 2) }
            satiety = if (rollChance(100) <= 40) {
                params[0] + randomizeParam()
            } else params[0]

            maxSatiety = if (rollChance(100) <= 40) {
                params[1] + randomizeParam()
            } else params[1]

            hungerThreshold = if (rollChance(100) < 40) {
                params[2] + randomizeParam()
            } else params[2]

            modificationInheritanceChance = if (rollChance(100) < 40) {
                params[3] + randomizeParam()
            } else params[3]
        }
    }

    fun addModification(modification: BacteriaModification) {
        if (modification.def != BacteriaModificationDef.NONE) {
            modifications.add(modification)
            children.add(modification)
        }

        modifications.removeIf { it.def == BacteriaModificationDef.NONE }
        satiety -= modification.def.price
        if (modification.def.outside) {
            modification.toBack()
        }
    }

    fun searchForFood(foodList: Set<NanoFoodPiece>) {
        if (satiety >= maxSatiety - 1) {
            MainLoop.printMessage("Bacteria is full")
        }

        val closestFood = foodList.minByOrNull { translatePosition.distance(it.translatePosition) }

        val tooFarAway = closestFood == null || closestFood.position.distance(translatePosition) > 1000

        if (tooFarAway) {
            return MainLoop.printMessage("Theres seems to be no food close enough")
        }

        foodTarget = closestFood ?: foodTarget

        caller = "fs"
        currentVelocity = baseVelocity
        moveToPoint(foodTarget.position)
    }

    private fun eatFoodTarget() {
        satiety = min(maxSatiety, satiety + (foodTarget.foodValue * 10).toFloat())
        foodTarget.eaten = true
        waitingTime = 120f
        state = OrganismState.WAITING
        caller = "idle"
    }

    fun searchForPartner(potentialPartners: Set<Bacteria>): Bacteria? {
        val closestBacteria = potentialPartners.filter { it != this }
            .maxByOrNull { it.position.distance(position) }

        closestBacteria ?:  return null

        val tooFarAway = closestBacteria.position.distance(position) > 1000

        if (tooFarAway) {
            return null
        }

        currentVelocity = baseVelocity
        return closestBacteria
    }

    fun makeMultiCellularOrganism(first: Bacteria, second: Bacteria?) {
        second ?: return MainLoop.printMessage("No other bacterias close enough")

        partnerTarget = second
        partnerTarget?.partnerTarget = first

        if (first.position.distance(second.position) > 50) {
            moveToPoint(first.position.midpoint(second.position))
            second.moveToPoint(second.position.midpoint(first.position))
            partnerTarget?.caller = "mcp"
            caller = "mcp"
        } else {
            bornMultiCellularOrganism()
        }
    }

    private fun bornMultiCellularOrganism() {
        partnerTarget ?: return
        MainLoop.bacteriaTarget ?: return
        val organism = MultiCellularOrganism()
        organism.translatePosition = MainLoop.bacteriaTarget!!.translatePosition
        MainLoop.multiCellularList.add(organism)
        MainLoop.objectRoot.children.add(organism)
        if (MainLoop.bacteriaTarget == this || MainLoop.bacteriaTarget == partnerTarget) {
            destroy()
            partnerTarget!!.destroy()
            MainLoop.bacteriaTarget = null
            MainLoop.multiCellularList += organism
        } else {
            destroy()
            partnerTarget!!.destroy()

        }
    }

    private fun destroy() {
        MainLoop.objectRoot.children.remove(this)
        remove = true
    }

    override fun toString() = buildString {
        append(gL("target")).append(":\n")
        append(gL("satiety")).append(" - ")
        append(round(satiety).toInt())
    }

    fun canMultiply(): Boolean  =  satiety >= 20f
}